pywren/wait.py

```diff
@@ -79,12 +110,57 @@ def _wait(fs, THREADPOOL_SIZE):
     # get the list of all objects in this callset
     callset_id = present_callsets.pop() # FIXME assume only one
 
-    storage_config = wrenconfig.extract_storage_config(wrenconfig.default())
-    storage_handler = storage.Storage(storage_config)
-    callids_done = storage_handler.get_callset_status(callset_id)
+    # note this returns everything done, so we have to figure out
+    # the intersection of those that are done
+    callids_done_in_callset = set(storage_handler.get_callset_status(callset_id))
+
+    not_done_call_ids = set([f.call_id for f in not_done_futures])
+
+    done_call_ids = not_done_call_ids.intersection(callids_done_in_callset)
+    not_done_call_ids = not_done_call_ids - done_call_ids
+
+    still_not_done_futures = [f for f in not_done_futures if (f.call_id in not_done_call_ids)]
 
-    callids_done = set(callids_done)
+    def fetch_future_status(f):
+        return storage_handler.get_call_status(f.callset_id, f.call_id)
 
+
+    pool = ThreadPool(THREADPOOL_SIZE)
+
+    # now try up to max_direct_query_n direct status queries, quitting once
+    # we have return_n done.
+    query_count = 0
+    max_queries = min(max_direct_query_n, len(still_not_done_futures))
+
+    if random_query:
+        random.shuffle(still_not_done_futures)
+
+    while query_count < max_queries:
+
+        if len(done_call_ids) >= return_early_n:
```

Comment:
I think with this, we assume that there are few stragglers? Let's say we have 100 stragglers (not showing up with list), then it will take  100/return_early_n * WAIT_DUR_SEC to finish. I personally have seen the case with a large number of stragglers yet, but just want to point out the possibility.

Reply:
Yeah I think in the future if we want and this becomes a performance pain point we can make return_early_n to be a user-facing parameter.